src/tests/test_numerictype_cast.cpp

```diff
@@ -0,0 +1,123 @@
+#include <memory>
+
+#include <gtest/gtest.h>
+
+#include "types.h"
+
+namespace tests {
+template <typename T>
+class TypePtr {};
+class NumericCast : public ::testing::Test {
+protected:
+    std::byte* dbl_a_val;
+    std::byte* dbl_b_val;
+    std::byte* int_a_val;
+    std::byte* int_b_val;
+
+    model::IntType int_type;
+    model::DoubleType dbl_type;
+    model::INumericType const& dbl_type_ref = dbl_type;
+    model::INumericType const& int_type_ref = int_type;
+
+    std::unique_ptr<std::byte[]> dbl_a_ptr;
+    std::unique_ptr<std::byte[]> dbl_b_ptr;
+    std::unique_ptr<std::byte[]> int_a_ptr;
+    std::unique_ptr<std::byte[]> int_b_ptr;
+
+    void SetUp() override {
+        int_a_ptr.reset(int_type_ref.MakeValueOfInt(21));
+        int_b_ptr.reset(int_type_ref.MakeValueOfInt(7));
+
+        int_a_val = int_a_ptr.get();
+        int_b_val = int_b_ptr.get();
+
+        dbl_a_ptr.reset(model::DoubleType::MakeFrom(int_a_val, int_type_ref));
+        dbl_b_ptr.reset(model::DoubleType::MakeFrom(int_b_val, int_type_ref));
```

Comment:
It's strange, I don't understand how memory leaks could be related to that, probably the problem was in something else

Reply:
It's strange, I don't understand how memory leaks could be related to that, probably the problem was in something else

Reply:
Method MakeValueOfInt calls MakeValue with same parameter as MakeValueOfInt, MakeValue allocates memory with operator new [], in case assertion fails test case will quit working and memory allocated with new[] will not be released. That was detected by clang-tidy-check warnings in begining of conversation

Reply:
I still don't understand. What's the difference between `int_seven_ptr.reset(int_type_ref.MakeValueOfInt(7));` and `dbl_seven_ptr.reset(dbl_type_ref.MakeValueOfInt(7));`? Why does the first one work fine while second one produce memory leaks?
Where this memory leak could possible come from if we're initializing unique pointer here?

Reply:
my mistake, have not got that we are speaking about double values (in first message there are INT_a/b_ptr, looked in code, got it, changing)

Reply:
this looks counterintuitive: MakeValueOfInt creates a double value